BibTeX.js

```diff
@@ -889,30 +890,67 @@ function beginRecord(type, closeChar) {
 }
 
 function doImport() {
-	var read = "", text = "", recordCloseElement = false;
+	if (typeof Promise == 'undefined') {
+		readString(
+			function () {},
+			function (e) {
+				throw e;
+			}
+		);
+	}
+	else {
+		return new Promise(function (resolve, reject) {
+			readString(resolve, reject);
+		});
+	}
+}
+
+function readString(resolve, reject) {
+	var read = "";
 	var type = false;
 	
-	while(read = Zotero.read(1)) {
-		if(read == "@") {
-			type = "";
-		} else if(type !== false) {
-			if(type == "comment") {
-				processComment();
-				type = false;
-			} else if(read == "{") {		// possible open character
-				beginRecord(type, "}");
-				type = false;
-			} else if(read == "(") {		// possible open character
-				beginRecord(type, ")");
-				type = false;
-			} else if(/[a-zA-Z0-9-_]/.test(read)) {
-				type += read;
+	var next = function () {
+		readString(resolve, reject);
+	};
+	
+	try {
+		while (read = Zotero.read(1)) {
+			if(read == "@") {
+				type = "";
+			} else if(type !== false) {
+				if(type == "comment") {
+					processComment();
+					type = false;
+				} else if(read == "{") {		// possible open character
+					// This might return a promise if an item was saved
+					// TODO: When 5.0-only, make sure this always returns a promise
+					let maybePromise = beginRecord(type, "}");
```

Comment:
Ugh, Safari 9. We'll be able to stop worrying about that (and hopefully start using various other features) after we drop support for 4.0, because 1) the new Safari connector requires macOS 10.11 anyway and 2) Zotero 5.0 requires 10.9, and all machines that can run 10.8 or later can be upgraded to 10.11. I think all other browser versions we care about support it (and have for a long time). But I've reverted these for now. Thanks.